q.ml

```diff
@@ -298,26 +298,78 @@ let to_float x =
          correctly-rounded result. *)
       Int64.to_float (Z.to_int64 p) /. Int64.to_float (Z.to_int64 q)
     else begin
-      (* |p| is in [2^(np-1), 2^np)
+      let negat =
+        if Z.lt p Z.zero then -1 else 1
+      in
+      (* p is in [2^(np-1), 2^np)
          q is in [2^(nq-1), 2^nq)
-         hence |p/q| is in (2^(np-nq-1), 2^(np-nq+1)).
-         We define n such that |p/q*2^n| is in [2^54, 2^56).
-         >= 2^54 so that the round to odd technique applies.
-         < 2^56 so that the integral part is representable as an int64. *)
-      let n = 55 - (np - nq) in
+         We define n,p',q' such that p'/q'*2^n=p/q and |p'/q'| is in [1, 2). *)
+      let n = np - nq in
       (* Scaling p/q by 2^n *)
       let (p', q') =
         if n >= 0
-        then (Z.shift_left p n, q)
-        else (p, Z.shift_left q (-n)) in
+        then p, Z.shift_left q n
+        else Z.shift_left p (-n), q
```

Comment:
One more style suggestion: add parentheses around these pairs like everywhere else in this file.
```suggestion
        then (p, Z.shift_left q n)
        else (Z.shift_left p (-n), q)
```